api/src/pcapi/core/providers/titelive_apply_new_filters_on_history.py

```diff
@@ -0,0 +1,348 @@
+PAPER_CONSUMABLE_SUPPORT_CODE = "PC"
+POSTER_SUPPORT_CODE = "PS"
+
+
+# This if the exact copy of bookings_api._cancel_bookings_from_stock with dry run support
+def _cancel_bookings_from_stock(
+    stock: models.Stock, reason: BookingCancellationReasons, dry: bool = False
+) -> list[bookings_models.Booking]:
+    """
+    Cancel multiple bookings and update the users' credit information on Batch.
+    Note that this will not reindex the stock.offer in Algolia
+    """
+    deleted_bookings: list[bookings_models.Booking] = []
+    for booking in stock.bookings:
+        if not dry and bookings_api._cancel_booking(
+            booking, reason, cancel_even_if_used=typing.cast(bool, stock.offer.isEvent)
+        ):
+            deleted_bookings.append(booking)
+        else:
+            deleted_bookings.append(
+                booking
+            )  # This doesn't exclude certain booking while in test mode because we don't handle exceptions
+    return deleted_bookings
+
+
+# This if the exact copy of bookings_api.cancel_bookings_from_rejected_offer with dry run support
+def cancel_bookings_from_rejected_offer(offer: models.Offer, dry: bool = False) -> list[bookings_models.Booking]:
+    cancelled_bookings = []
+    for stock in offer.stocks:
+        cancelled_bookings.extend(_cancel_bookings_from_stock(stock, BookingCancellationReasons.FRAUD, dry=dry))
+    logger.info(
+        "Cancelled bookings for rejected offer",
+        extra={
+            "bookings": [b.id for b in cancelled_bookings],
+            "offer": offer.id,
+        },
+    )
+
+    return cancelled_bookings
+
+
+# This if the exact copy of offers_api.reject_inappropriate_products with dry run support and return list of email instead of bool
+def reject_inappropriate_products(
+    ean: str, send_booking_cancellation_emails: bool = True, dry: bool = False
+) -> list[str]:
+    products = models.Product.query.filter(models.Product.extraData["ean"].astext == ean).all()
+    emails: list[str] = []
+
+    if not products:
+        return emails
+
+    for product in products:
+        product.isGcuCompatible = False
+        db.session.add(product)
+
+    offers_query = models.Offer.query.filter(
+        models.Offer.productId.in_(p.id for p in products),
+        models.Offer.validation != models.OfferValidationStatus.REJECTED,
+    ).options(sa.orm.joinedload(models.Offer.stocks).joinedload(models.Stock.bookings))
+
+    offers = offers_query.all()
+
+    if not dry:
+        offers_query.update(
+            values={
+                "validation": models.OfferValidationStatus.REJECTED,
+                "lastValidationDate": datetime.datetime.utcnow(),
+                "lastValidationType": OfferValidationType.CGU_INCOMPATIBLE_PRODUCT,
+            },
+            synchronize_session="fetch",
+        )
+
+    offer_ids = []
+    for offer in offers:
+        offer_ids.append(offer.id)
+        bookings = cancel_bookings_from_rejected_offer(offer, dry=dry)
```

Comment:
Je trouve que ça complique le script de redéfinir `cancel_bookings_from_rejected_offer()` et `_cancel_bookings_from_stock()`. Sans compter que si jamais quelque chose change dans la vraie implémentation de ces méthodes, ton script ne sera pas à jour. J'ai l'impression qu'on peut simplifier ça en écrivant comme suit : 

```
if dry:
    bookings = list(itertools.chain.from_iterable(stock.bookings for stock in offer.stocks))
else:
    bookings = bookings_api.cancel_bookings_from_rejected_offer(offer)
```

Je loupe quelque chose ?

Reply:
à chaque modification par rapport au script de production, j'ai peur de m'écarter de la réalité. En plus, ta ligne est très différentes de ce que je lis. 

Du coup, je sais pas te répondre pour si le code est pareil, mais est-ce vraiment une bonne idée ?

Reply:
Ça me semble au contraire beaucoup plus simple de lire ma version, plutôt que de plonger dans les fonctions que tu as redéfinies, supposer que ta version ne diverge pas trop des vraies implémentations dans `core/bookings/api.py` et constater que, finalement, en mode "dry run", on ne fait aucune modification et on se contente re-renvoyer la liste des réservations. Non ?

Reply:
Je ne sais toujours pas, est-ce possible d'en parler à l'oral ?

Reply:
C'est ce qu'il se passait dans `bookings_api.cancel_bookings_from_rejected_offer` qui me parraissait compliqué a lire et a affirmer quie t'as traduction en `list(itertools.chain.from_iterable(stock.bookings for stock in offer.stocks))`  soit bonne. Mais je te fais confiance et j'ai maj la PR